File: suites/12_game_ai/comparisons/minimax_search.go
package main

import "fmt"

func evaluateBoard(state int64) int64 {
	score := int64(0)
	for i := int64(0); i < 9; i++ {
		piece := (state >> (i * 2)) & 3
		if piece == 1 { score += 10 }
		if piece == 2 { score -= 10 }
	}
	return score
}

func minimax(state, depth int64, isMax bool) int64 {
	if depth == 0 { return evaluateBoard(state) }

	if isMax {
		bestVal := int64(-999999)
		for i := int64(0); i < 9; i++ {
			if ((state >> (i * 2)) & 3) == 0 {
				nextState := state | (1 << (i * 2))
				val := minimax(nextState, depth-1, false)
				if val > bestVal { bestVal = val }
			}
		}
		if bestVal == -999999 { return evaluateBoard(state) }
		return bestVal
	} else {
		bestVal := int64(999999)
		for i := int64(0); i < 9; i++ {
			if ((state >> (i * 2)) & 3) == 0 {
				nextState := state | (2 << (i * 2))
				val := minimax(nextState, depth-1, true)
				if val < bestVal { bestVal = val }
			}
		}
		if bestVal == 999999 { return evaluateBoard(state) }
		return bestVal
	}
}

func minimaxSearch(games int64) int64 {
	checksum := int64(0)
	for g := int64(0); g < games; g++ {
		initialState := (g * 12345) % 262144
		result := minimax(initialState, 6, true)
		checksum = (checksum*31 + result) % 1000000007
	}
	return checksum
}

func main() { fmt.Println(minimaxSearch(64)) }

File: suites/12_game_ai/comparisons/minimax_search.rs
use std::cmp::{max, min};

fn evaluate_board(state: i64) -> i64 {
    let mut score: i64 = 0;
    for i in 0..9 {
        let piece = (state >> (i * 2)) & 3;
        if piece == 1 { score += 10; }
        if piece == 2 { score -= 10; }
    }
    score
}

fn minimax(state: i64, depth: i64, is_max: bool) -> i64 {
    if depth == 0 { return evaluate_board(state); }
    
    if is_max {
        let mut best_val = -999999;
        for i in 0..9 {
            if ((state >> (i * 2)) & 3) == 0 {
                let next_state = state | (1 << (i * 2));
                let val = minimax(next_state, depth - 1, false);
                best_val = max(best_val, val);
            }
        }
        if best_val == -999999 { evaluate_board(state) } else { best_val }
    } else {
        let mut best_val = 999999;
        for i in 0..9 {
            if ((state >> (i * 2)) & 3) == 0 {
                let next_state = state | (2 << (i * 2));
                let val = minimax(next_state, depth - 1, true);
                best_val = min(best_val, val);
            }
        }
        if best_val == 999999 { evaluate_board(state) } else { best_val }
    }
}

fn minimax_search(games: i64) -> i64 {
    let mut checksum: i64 = 0;
    for g in 0..games {
        let initial_state = (g * 12345) % 262144;
        let result = minimax(initial_state, 6, true);
        checksum = (checksum * 31 + result) % 1000000007;
    }
    checksum
}

fn main() { println!("{}", minimax_search(64)); }
